## Importando base do Kaggle.

Kaggle é uma plataforma que une a comunidade de cientista de dados, nele temos alguns Datasets que podemos usar para fins de experimento de um modelo ou para um propósito acadêmico por exemplo.

https://www.kaggle.com/

In [2]:
import pandas as pd

BASE = "../datasets/Pokemon.csv" # https://www.kaggle.com/datasets/jaidalmotra/pokemon-dataset

About Dataset
This data set includes 898 Pokemon, 1072 including alternate forms, including their number, name, first and second type, the stat total and basic stats: HP, Attack, Defense, Special Attack, Special Defense, and Speed, generation, and legendary status. It has been of great use when teaching statistics to kids. With certain types you can also give a geeky introduction to machine learning.

These are the raw attributes that are used for calculating how much damage an attack will do in the games. This dataset is about the pokemon games (NOT pokemon cards or Pokemon Go).

Number: The ID for each pokemon

Name: The name of each pokemon

Type 1: Each pokemon has a type, this determines weakness/resistance to attacks

Type 2: Some pokemon are dual type and have 2

Total: Sum of all stats that come after this, a general guide to how strong a pokemon is

HP: Hit points, or health, defines how much damage a pokemon can withstand before fainting

Attack: The base modifier for normal attacks (eg. Scratch, Punch)

Defense: The base damage resistance against normal attacks

SP Atk: Special attack, the base modifier for special attacks (e.g. fire blast, bubble beam)

SP Def: Special defense, the base damage resistance against special attacks

Speed: Determines which pokemon attacks first each round

Generation: The generation of games where the pokemon was first introduced

Legendary: Some pokemon are much rarer than others, and are dubbed "legendary"

Inspiration: The type of a pokemon cannot be inferred only by its Attack and Defense. It would be worthy to find which two variables can define the type of a pokemon, if any. Two variables can be plotted in a 2D space, and used as an example for machine learning. This could mean the creation of a visual example any geeky Machine Learning class would love.


View less
Usability


In [3]:
df = pd.read_csv(BASE)

##  .head()

Por padrão o .head exibe as 5 primeiras linhas do nosso DataFrame, mas se quisermos ver algum valor diferente de linhas podemos passar isso como argumento para a função.

In [16]:
df.head(10)

number                 name  type1   type2  total  hp  attack  defense  \
0       1            Bulbasaur  Grass  Poison    318  45      49       49   
1       2              Ivysaur  Grass  Poison    405  60      62       63   
2       3             Venusaur  Grass  Poison    525  80      82       83   
3       3        Mega Venusaur  Grass  Poison    625  80     100      123   
4       3  Gigantamax Venusaur  Grass  Poison    525  80      82       83   
5       4           Charmander   Fire     NaN    309  39      52       43   
6       5           Charmeleon   Fire     NaN    405  58      64       58   
7       6            Charizard   Fire  Flying    534  78      84       78   
8       6     Mega Charizard X   Fire  Dragon    634  78     130      111   
9       6     Mega Charizard Y   Fire  Flying    634  78     104       78   

   sp_attack  sp_defense  speed  generation  legendary  
0         65          65     45           1      False  
1         80          80     60           1      False  
2        100         100     80           1      False  
3        122         120     80           1      False  
4        100         100     80           1      False  
5         60          50     65           1      False  
6         80          65     80           1      False  
7        109          85    100           1      False  
8        130          85    100           1      False  
9        159         115    100           1      False

##  .tail()

O .tail() tem o mesmo propósito do .head() a diferença é que ele exibirá os 5 ultimos registros do nosso DataFrame, mas podemos passar um valor customizado pra ele também.

In [18]:
df.tail(10)

number                                   name     type1  type2  total  \
1062     892  Gigantamax Urshifu Rapid Strike Style  Fighting  Water    550   
1063     893                                 Zarude      Dark  Grass    600   
1064     893                            Dada Zarude      Dark  Grass    600   
1065     894                              Regieleki  Electric    NaN    580   
1066     895                              Regidrago    Dragon    NaN    580   
1067     896                              Glastrier       Ice    NaN    580   
1068     897                              Spectrier     Ghost    NaN    580   
1069     898                                Calyrex   Psychic  Grass    500   
1070     898                      Ice Rider Calyrex   Psychic    Ice    680   
1071     898                   Shadow Rider Calyrex   Psychic  Ghost    680   

       hp  attack  defense  sp_attack  sp_defense  speed  generation  \
1062  100     130      100         63          60     97           8   
1063  105     120      105         70          95    105           8   
1064  105     120      105         70          95    105           8   
1065   80     100       50        100          50    200           8   
1066  200     100       50        100          50     80           8   
1067  100     145      130         65         110     30           8   
1068  100      65       60        145          80    130           8   
1069  100      80       80         80          80     80           8   
1070  100     165      150         85         130     50           8   
1071  100      85       80        165         100    150           8   

      legendary  
1062       True  
1063       True  
1064       True  
1065       True  
1066       True  
1067       True  
1068       True  
1069       True  
1070       True  
1071       True

##  .shape

O .shape é um atributo do nosso DataFrame que nos informa em forma de tupla o total de linhas e colunas respectivamente.

In [7]:
df.shape

(1072, 13)

##  .columns

O .columns é um atributo do nosso DataFrame que nos informa em forma de array de strings os nomes das nossas colunas.

In [12]:
df.columns

Index(['number', 'name', 'type1', 'type2', 'total', 'hp', 'attack', 'defense',
       'sp_attack', 'sp_defense', 'speed', 'generation', 'legendary'],
      dtype='object')

## O .isnull() com o .sum()

O .isnull nos retorna o nosso DataFrame informando para cada célula se aquele valor é Null ou não em forma de Booleano (True se for Null e False em caso contrário)

In [13]:
df.isnull()

number   name  type1  type2  total     hp  attack  defense  sp_attack  \
0      False  False  False  False  False  False   False    False      False   
1      False  False  False  False  False  False   False    False      False   
2      False  False  False  False  False  False   False    False      False   
3      False  False  False  False  False  False   False    False      False   
4      False  False  False  False  False  False   False    False      False   
...      ...    ...    ...    ...    ...    ...     ...      ...        ...   
1067   False  False  False   True  False  False   False    False      False   
1068   False  False  False   True  False  False   False    False      False   
1069   False  False  False  False  False  False   False    False      False   
1070   False  False  False  False  False  False   False    False      False   
1071   False  False  False  False  False  False   False    False      False   

      sp_defense  speed  generation  legendary  
0          False  False       False      False  
1          False  False       False      False  
2          False  False       False      False  
3          False  False       False      False  
4          False  False       False      False  
...          ...    ...         ...        ...  
1067       False  False       False      False  
1068       False  False       False      False  
1069       False  False       False      False  
1070       False  False       False      False  
1071       False  False       False      False  

[1072 rows x 13 columns]

Mas não é tão comum visualizarmos nesse formato a comunidade de ciência de dados costuma utilizá-lo com o .sum() que vai nos informar o total de registros nulos por coluna.

In [14]:
df.isnull().sum()

number          0
name            0
type1           0
type2         498
total           0
hp              0
attack          0
defense         0
sp_attack       0
sp_defense      0
speed           0
generation      0
legendary       0
dtype: int64

## Usando o `with` para otimizar a visualização do .describe()

O with serve pra 2 propósitos - Finalizar recusrsos adquiridos, ou seja, garantir que uma operação que nosso programa iniciou seja finalizada. E serve também para contextualizar apenas em um determinado bloco nossa alteração, ou seja, o que for manipulado internamente em um `.with` só vai ter efeito dentro dele. 

In [23]:
with pd.option_context("float_format", "{:.2f}".format): # O que estiver dentro deste with terá os valores float ajustados
    display(df.describe())

# df.describe()  Este trecho não terá os valores float formatados

number   total      hp  attack  defense  sp_attack  sp_defense   speed  \
count 1072.00 1072.00 1072.00 1072.00  1072.00    1072.00     1072.00 1072.00   
mean   445.22  440.89   70.49   80.94    74.97      73.27       72.48   68.79   
std    267.77  121.38   26.87   32.46    31.21      32.64       27.93   30.08   
min      1.00  175.00    1.00    5.00     5.00      10.00       20.00    5.00   
25%    209.75  330.00   50.00   56.00    52.00      50.00       50.00   45.00   
50%    442.50  460.50   68.00   80.00    70.00      65.00       70.00   65.00   
75%    681.25  519.25   84.00  100.00    90.00      95.00       90.00   90.00   
max    898.00 1125.00  255.00  190.00   250.00     194.00      250.00  200.00   

       generation  
count     1072.00  
mean         4.29  
std          2.35  
min          0.00  
25%          2.00  
50%          4.00  
75%          6.00  
max          8.00

In [24]:
df.describe(exclude="number") # visualizando as colunas não numéricas do DataSet.

name  type1   type2 legendary
count        1072   1072     574      1072
unique       1072     20      18         2
top     Bulbasaur  Water  Flying     False
freq            1    136     114       954

O DataFrame com as informações sobre nosso DataSet nos diz que: 
 - Há 1072 registros das colunas name, type1 e legendary e 574 registros na coluna type2
 - 1072 registros unicos da coluna name 20 registros unicos da coluna type1, 18 da coluna type2 e 2 na coluna legendary
 - Olhando as linhas  top e freq, podemos concluir que na coluna name o registro que mais se repete é o  "Bulbasaur" que segundo a linha freq é exibido uma vez, já na coluna type1 o registro que mais aparece é "Water" que é exibido 136 vezes, na coluna type2 o registro que é mais exibido é "Flying" 114 vezes - E por ultimo temos a coluna legendary que exibe 954 vezes o registro "False".

## .nunique()

Nos retorna o total de registros unicos por coluna.

In [25]:
df.nunique()

number         898
name          1072
type1           20
type2           18
total          219
hp             103
attack         123
defense        114
sp_attack      119
sp_defense     107
speed          127
generation       9
legendary        2
dtype: int64

## Usando o select_dtypes para obter apenas colunas  de determinado tipo

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.select_dtypes.html

In [27]:
df.select_dtypes("object") # Retorna apenas colunas do tipo object do nosso DataFrame

name    type1   type2
0                Bulbasaur    Grass  Poison
1                  Ivysaur    Grass  Poison
2                 Venusaur    Grass  Poison
3            Mega Venusaur    Grass  Poison
4      Gigantamax Venusaur    Grass  Poison
...                    ...      ...     ...
1067             Glastrier      Ice     NaN
1068             Spectrier    Ghost     NaN
1069               Calyrex  Psychic   Grass
1070     Ice Rider Calyrex  Psychic     Ice
1071  Shadow Rider Calyrex  Psychic   Ghost

[1072 rows x 3 columns]

## Obtendo os registros unicos por coluna

In [33]:
df['type1'].unique()

array(['Grass', 'Fire', 'Water', 'Blastoise', 'Bug', 'Normal', 'Dark',
       'Poison', 'Electric', 'Ground', 'Ice', 'Fairy', 'Steel',
       'Fighting', 'Psychic', 'Rock', 'Ghost', 'Dragon', 'Flying',
       'Graass'], dtype=object)